libbpf-tools/schedsnoop.c

```diff
@@ -0,0 +1,309 @@
+		siv = ti->ts;
+		sik.cpu = ti->cpu;
+		sik.pid = ti->pid;
+		sik.syscall = ti->syscall;
+		bpf_map_update_elem(si_map_fd, &sik, &siv, BPF_ANY);
+		syscall_name(ti->syscall, comm_buf, sizeof(comm_buf));
+		snprintf(func, sizeof(func), "SC [%d:%s] ENTER",
+				ti->syscall, comm_buf);
+		pr_ti(ti, func, NULL);
+		break;
+	case TYPE_SYSCALL_EXIT:
+		sik.cpu = ti->cpu;
+		sik.pid = ti->pid;
+		sik.syscall = ti->syscall;
+		if (bpf_map_lookup_elem(si_map_fd, &sik, &siv))
+			break;
+		time_to_str(ti->ts - siv, d_str, sizeof(d_str));
+		bpf_map_delete_elem(si_map_fd, &sik);
+		syscall_name(ti->syscall, comm_buf, sizeof(comm_buf));
+		snprintf(func, sizeof(func), "SC [%d:%s] TAKE %s TO EXIT",
+				ti->syscall, comm_buf, d_str);
+		pr_ti(ti, func, NULL);
+		break;
+	default:
+		break;
+	}
+
+	last_time = ti->ts;
+}
+
+void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	fprintf(stderr, "Lost %llu events on CPU #%d!\n", lost_cnt, cpu);
+}
+
+static void int_exit(int sig)
+{
+	exiting = true;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct perf_buffer_opts pb_opts;
+	struct perf_buffer *pb = NULL;
+	struct schedsnoop_bpf *obj;	
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	init_syscall_names();
+	
+	/* Check cpu number */
+	env.nr_cpus = sysconf(_SC_NPROCESSORS_CONF);
+	if (env.nr_cpus > NR_CPU_MAX) {
+		printf("Support Maximum %d cpus\n", NR_CPU_MAX);
+		goto freename;
+	}
+	
+	/* Increase rlimit */
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		goto freename;
+	}
+	
+	/* Open bpf object */
+	obj = schedsnoop_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		goto freename;
+	}
+
+	/* initialize global data (filtering options) */
+	obj->rodata->targ_pid = env.target;
+	obj->rodata->trace_syscall = env.trace_syscall;
+	
+	/* Load bpf program */
+	err = schedsnoop_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+	
+	/* Attach bpf program */
+	err = schedsnoop_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+	
+	/* Setup global map fd */
+	si_map_fd = bpf_map__fd(obj->maps.syscall_info_maps);
+	
+	signal(SIGINT, int_exit);
+	signal(SIGTERM, int_exit);
+
+	printf("Start tracing schedule events ");
+	if (env.trace_syscall)
+		printf("(include SYSCALL)");
+	printf("\nTarget task pid %d\n", env.target);
+	
+	/* setup event callbacks */
+	pb_opts.sample_cb = handle_event;
+	pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES, &pb_opts);
+	err = libbpf_get_error(pb);
+	if (err) {
+		pb = NULL;
+		fprintf(stderr, "failed to open perf buffer: %d\n", err);
+		goto cleanup;
+	}
+	
+	/* main: poll */
+	while (!exiting && !obj->bss->targ_exit && \

```

Comment:
hmm, you don't need '\' in C statements, I'm even surprised that compiles (but I guess it makes sense). Still, quite unconventional and unnecessary.